src/Numerics/SystemSolvers/batched_generalized_minimal_residual_solver.jl

```diff
@@ -263,517 +276,381 @@ function BatchedGeneralizedMinimalResidual(
         max_subspace_size = m
     end
 
-    # Now we need to determine an appropriate permutation
-    # of the MPIStateArray to perform column-wise strides.
-    # Total size of the permute tuple
+    # permute [ni, nj, nk, num_states, nvertelem, nhorzelem] 
+    # to      [nvertelem, nk, num_states, ni, nj, nhorzelem]
     permute_size = length(reshaping_tup)
-
-    # Index associated with number of GL points
-    # in the 'vertical' direction
-    nql = length(Np)
-    # Want: (index associated with the stack size of the column,
-    #        index associated with GL pts in vertical direction,
-    #        index associated with number of states)
-    # FIXME: Better way to do this?
-    #             (vert stack, GL pts, num states)
-    column_strides = (nql + 2, nql, nql + 1)
-    diff = Tuple(setdiff(Set([i for i in 1:permute_size]), Set(column_strides)))
-    permute_tuple_f = (column_strides..., diff...)
+    permute_tuple_f = (dim + 2, dim, dim + 1, (1:(dim - 1))..., permute_size)
 
     return BatchedGeneralizedMinimalResidual(
-        Q;
-        m = m,
-        n = n,
-        subspace_size = max_subspace_size,
+        Q,
+        m,
+        n;
+        M = max_subspace_size,
         atol = atol,
         rtol = rtol,
-        ArrayType = ArrayType,
-        reshape_tuple_f = reshaping_tup,
-        permute_tuple_f = permute_tuple_f,
+        forward_reshape = reshaping_tup,
+        forward_permute = permute_tuple_f,
     )
 end
 
-# initialize function (1)
 function initialize!(
     linearoperator!,
     Q,
     Qrhs,
     solver::BatchedGeneralizedMinimalResidual,
-    args...,
+    args...;
+    restart = false,
 )
-    # body of initialize function in abstract iterative solver
-    return false, zero(eltype(Q))
+    g0 = solver.g0
+    krylov_basis = solver.krylov_basis
+    rtol, atol = solver.rtol, solver.atol
+
+    batched_krylov_basis = solver.batched_krylov_basis
+    forward_reshape = solver.forward_reshape
+    forward_permute = solver.forward_permute
+    resnorms = solver.resnorms
+    resnorms0 = solver.resnorms0
+
+    # Device and groupsize information
+    device = array_device(Q)
+    groupsize = 256
+
+    @assert size(Q) == size(krylov_basis)
+
+    # FIXME: Can we make linearoperator! batch-able?
+    # store the initial (global) residual in krylov_basis = r0/|r0|
+    # solve for w then apply
+    # PRECONDITIONER: Q ->  P^{-1}Q
+    linearoperator!(krylov_basis, Q, args...)
+    krylov_basis .= Qrhs .- krylov_basis
+
+    # Convert into a batched Krylov basis vector
+    # REMARK: Ugly hack on the GPU. Can we fix this?
+    tmp_array = similar(batched_krylov_basis[1, :, :])
```

Comment:
```suggestion
    tmp_array = similar(batched_krylov_basis, size(batched_krylov_basis)[2:3])
```
No need to create a slice and then another array

Reply:
Done!